Collector/RequestStackProvider.php

```diff
@@ -0,0 +1,141 @@
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class RequestStackProvider
+{
+    /**
+     * Array that tell if a request errored or not. true = success, false = failure.
+     *
+     * @var array
+     */
+    private $failure;
+
+    /**
+     * A multidimensional array with requests.
+     * $requests[0][0] is the first request before all plugins.
+     * $requests[0][1] is the first request after the first plugin.
+     *
+     * @var array
+     */
+    private $requests;
+
+    /**
+     * A multidimensional array with responses.
+     * $responses[0][0] is the first responses before all plugins.
+     * $responses[0][1] is the first responses after the first plugin.
+     *
+     * @var array
+     */
+    private $responses;
+
+    /**
+     * @param array $failure   if the response was successful or not
+     * @param array $requests
+     * @param array $responses
+     */
+    public function __construct(array $failure, array $requests, array $responses)
+    {
+        $this->failure = $failure;
+        $this->requests = $requests;
+        $this->responses = $responses;
+    }
+
+    /**
+     * Create an array of ClientDataCollector from collected data.
+     *
+     * @param array $data
+     *
+     * @return RequestStackProvider[]
+     */
+    public static function createFromCollectedData(array $data)
+    {
+        $clientData = [];
+        foreach ($data as $clientName => $messages) {
+            $clientData[$clientName] = static::createOne($messages);
+        }
+
+        return $clientData;
+    }
+
+    /**
+     * @param array $messages is an array with keys 'failure', 'request' and 'response' which hold requests for each call to
+     *                        sendRequest and for each depth
+     *
+     * @return RequestStackProvider
+     */
+    private static function createOne($messages)
+    {
+        $orderedFaulure = [];
+        $orderedRequests = [];
+        $orderedResponses = [];
+
+        foreach ($messages['failure'] as $depth => $failures) {
+            foreach ($failures as $idx => $failure) {
+                $orderedFaulure[$idx][$depth] = $failure;
+            }
+        }
+
+        foreach ($messages['request'] as $depth => $requests) {
+            foreach ($requests as $idx => $request) {
+                $orderedRequests[$idx][$depth] = $request;
+            }
+        }
+
+        foreach ($messages['response'] as $depth => $responses) {
+            foreach ($responses as $idx => $response) {
+                $orderedResponses[$idx][$depth] = $response;
+            }
+        }
+
+        return new self($orderedFaulure, $orderedRequests, $orderedResponses);
+    }
+
+    /**
+     * Get the index keys for the request and response stacks.
+     *
+     * @return array
+     */
+    public function getStackIndexKeys()
+    {
+        return array_keys($this->requests);
+    }
+
+    /**
+     * @param int $idx
+     *
+     * @return array responses
```

Comment:
Not sure if we need these descriptions here. It is not verbose enough and can be confusing. Either have little but more verbose description or none IMO. Same for below.